### PR TITLE
Sets - Insert or update

### PR DESCRIPTION
**Pull request by clementetb**
Adds support for the `insertOrUpdate` operation on objects/object collections containing sets

### CODE REVIEW

File: realm/realm-annotations-processor/src/main/java/io/realm/processor/RealmProxyClassGenerator.kt [hunk 1]
```diff
@@ -1921,9 +1921,85 @@ class RealmProxyClassGenerator(private val processingEnvironment: ProcessingEnvi
...
-                        emitSingleLineComment("TODO: Set")
+                    Utils.isRealmModelSet(field) -> {
+                        val genericType: TypeMirror = Utils.getGenericType(field)!!
+                        val isEmbedded = Utils.isFieldTypeEmbedded(genericType, classCollection)
+
+                        emitEmptyLine()
+                        emitStatement("OsSet %1\$sOsSet = new OsSet(table.getUncheckedRow(objKey), columnInfo.%1\$sColKey)", fieldName)
+                        emitStatement("RealmSet<%s> %sSet = ((%s) object).%s()", genericType, fieldName, interfaceName, getter)
+                        if (isEmbedded) {
+                            // throw not supported
```
Comment (edualonso) [inline realm/realm-annotations-processor/src/main/java/io/realm/processor/RealmProxyClassGenerator.kt:1932]: Not sure I understand this comment. Also, I cannot remember the full implementation details, but should we not throw and cancel compilation if the type is an embedded object? Or do we handle that case in runtime?

File: realm/realm-library/src/main/java/io/realm/RealmAnyOperator.java [hunk 2]
```diff
@@ -406,7 +406,7 @@ protected NativeRealmAny createNativeRealmAny() {
 
     @Override
     Class<?> getTypedClass() {
-        return clazz;
+        return RealmObjectProxy.class.isAssignableFrom(clazz) ? clazz.getSuperclass() : clazz;
```
Comment (edualonso) [inline realm/realm-library/src/main/java/io/realm/RealmAnyOperator.java:409]: Why this change?

**Review by edualonso — COMMENTED** [review]
Just two comments.

File: realm/realm-annotations-processor/src/main/java/io/realm/processor/RealmProxyClassGenerator.kt (hunk 1 above)
Comment (clementetb) [reply realm/realm-annotations-processor/src/main/java/io/realm/processor/RealmProxyClassGenerator.kt:1932]: Forgot to add the piece of code that handles the case.

File: realm/realm-library/src/main/java/io/realm/RealmAnyOperator.java (hunk 2 above)
Comment (clementetb) [reply realm/realm-library/src/main/java/io/realm/RealmAnyOperator.java:409]: It is a bug fix. If you initialize a RealmAny with a Managed object it would return the Proxy class as the value class instead of the base model class. It is an issue when you try to compare the value class against a model class like it is done in the module mediator.

I am adding a test case to validate the behavior.

**Review by edualonso — APPROVED** [review]
